app/src/main/java/org/schabi/newpipe/settings/SettingMigrations.java

```diff
@@ -0,0 +1,131 @@
+
+public final class SettingMigrations {
+    /**
+     * Version number for preferences. Must be incremented every time a migration is necessary.
+     */
+    public static final int VERSION = 2;
+    private static SharedPreferences sp;
+
+    public static final Migration MIGRATION_0_1 = new Migration(0, 1) {
+        @Override
+        public void migrate(final Context context) {
+            // We changed the content of the dialog which opens when sharing a link to NewPipe
+            // by removing the "open detail page" option.
+            // Therefore, show the dialog once again to ensure users need to choose again and are
+            // aware of the changed dialog.
+            final SharedPreferences.Editor editor = sp.edit();
+            editor.putString(context.getString(R.string.preferred_open_action_key),
+                    context.getString(R.string.always_ask_open_action_key));
+            editor.apply();
+        }
+    };
+
+    public static final Migration MIGRATION_1_2 = new Migration(1, 2) {
+        @Override
+        protected void migrate(final Context context) {
+            // The new application workflow introduced in #2907 allows minimizing videos
+            // while playing to do other stuff within the app.
+            // For an even better workflow, we minimize a stream when switching the app to play in
+            // background.
+            // Therefore, set default value to background, if it has not been changed yet.
+            final String minimizeOnExitKey = context.getString(R.string.minimize_on_exit_key);
+            if (sp.getString(minimizeOnExitKey, "")
+                    .equals(context.getString(R.string.minimize_on_exit_none_key))) {
+                final SharedPreferences.Editor editor = sp.edit();
+                editor.putString(minimizeOnExitKey,
+                        context.getString(R.string.minimize_on_exit_background_key));
+                editor.apply();
+            }
+        }
+    };
+
+    /**
+     * List of all implemented migrations.
+     * <p>
+     * <b>Append new migrations to the end of the list</b> to keep it sorted ascending.
+     * If not sorted correctly, migrations which depend on each other, may fail.
+     */
+    private static final Migration[] SETTING_MIGRATIONS = {
+            MIGRATION_0_1,
+            MIGRATION_1_2
+    };
+
+
+    public static void initMigrations(final Context context, final boolean isFirstRun) {
+        // setup migrations and check if there is something to do
+        sp = PreferenceManager.getDefaultSharedPreferences(context);
+        final String lastPrefVersionKey = context.getString(R.string.last_used_preferences_version);
+        final int lastPrefVersion = sp.getInt(lastPrefVersionKey, 0);
+
+        // no migration to run, already up to date
+        if (isFirstRun) {
+            sp.edit().putInt(lastPrefVersionKey, VERSION).apply();
+            return;
+        } else if (lastPrefVersion == VERSION) {
```

Comment:
Isn't VERSION equal to SETTINGS_MIGRATIONS[-1].newVersion? Even though fixing this would probably reduce redundancy, I am not sure if it should be done, since what you have done looks clearer

Reply:
That's correct, but as you said, this makes the code harder to read.